### PR TITLE
Sup-8515 :  Adding Truststore config to read certs from NSI

### PR DESCRIPTION
**Pull request by sureshhmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala [hunk 1]
```diff
@@ -84,31 +75,48 @@ class CustomWSConfigParser @Inject()(configuration: Configuration, env: Environm
...
+        val keyStore = KeyStore.getInstance(KeyStore.getDefaultType)
+        keyStore.load(null, null)
+
+        val certs = generateCertificates(trustFile)
+        certs.foreach { cert ⇒
+          val alias = cert.asInstanceOf[X509Certificate].getSubjectX500Principal.getName
+          keyStore.setCertificateEntry(alias, cert)
+        }
+
+        keyStore.store(new FileOutputStream(trustFile.getAbsolutePath), "".toCharArray)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala]: Do we not need to call `close` on the `FileOutputStream` when we've finished using it for efficiency?

File: app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala [hunk 2]
```diff
@@ -84,31 +75,48 @@ class CustomWSConfigParser @Inject()(configuration: Configuration, env: Environm
...
+      case Failure(error) ⇒
+        logger.error(s"Error storing trust data in temp file", error)
+        sys.error(s"Error storing trust data in temp file: ${error.getMessage}")
+    }
   }
 
+  private def generateCertificates(file: File): Seq[Certificate] = {
 
-  def writeToTempFile(data: Array[Byte]): Try[File] = Try {
+    val dis = new DataInputStream(new FileInputStream(file))
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala]: we should probably be closing these streams as well

File: app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala [hunk 3]
```diff
@@ -84,31 +75,48 @@ class CustomWSConfigParser @Inject()(configuration: Configuration, env: Environm
...
+    val bytes = new Array[Byte](dis.available)
+    dis.readFully(bytes)
+    val bais = new ByteArrayInputStream(bytes)
+    val cf = CertificateFactory.getInstance("X.509")
+    cf.generateCertificates(bais).toList
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala]: Doesn't
```scala
private def generateCertificates(file: File): Seq[Certificate] = {
  val fileInputStream = new FileInputStream(file)
  val factory = CertificateFactory.getInstance("X.509")
  try {
    cf.generateCertificates(bais).toList
  } finally {
   case NonFatal(_) =>
      fileInputStream.close()
     // handle error (sys.error ?)
  }
}
```
work? Do we need to create the `ByteArrayInputStream`? Also, I made sure that the resources were closed even in the case of an exception being thrown

File: conf/application.conf [hunk 4]
```diff
@@ -46,7 +49,7 @@ play.ws.ssl {
     ssl = false
     sslctx = false
     defaultctx = false
-    handshake = true
+    handshake = false
```
Comment (AndyHWChung) [inline conf/application.conf]: why is this `false`?

File: test/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnectorSpec.scala [hunk 5]
```diff
@@ -48,7 +48,7 @@ class NSIConnectorSpec extends TestSupport with MockFactory {
     (mockHTTPProxy.post(
       _: String, _: I, _: Map[String, String]
     )(_: Writes[I], _: HeaderCarrier))
-      .expects(nsiUrl, body, Map(nsiAuthHeaderKey → nsiBasicAuth), *, hc.copy(authorization = None))
+      .expects(nsiUrl, body, Map(nsiAuthHeaderKey → nsiBasicAuth), *, *)
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnectorSpec.scala]: could have `hc.copy(token = None)` now

File: app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala [hunk 6]
```diff
@@ -84,31 +75,48 @@ class CustomWSConfigParser @Inject()(configuration: Configuration, env: Environm
...
-      config.copy(underlying = config.underlying.withValue(s"play.ws.ssl.${name}Manager.stores", ConfigValueFactory.fromIterable(stores)))
-    } else config
+  private def createTrustStoreConfig(ts: TrustStoreConfig, data: String): TrustStoreConfig = {
+
+    createTempFileForData(data) match {
+      case Success(trustFile) ⇒
+        val keyStore = KeyStore.getInstance(KeyStore.getDefaultType)
+        keyStore.load(null, null)
+
+        val certs = generateCertificates(trustFile)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/config/CustomWSConfigParser.scala]: could we not make

 `def createTempFileForData(data: String): Try[File]`

not have the type signature

` def createTempFileForData(data: String): Array[Byte]`

instead?  Then `generateCertificates` could be implemented like this:
```scala
private def generateCertificates(file: Array[Byte]): Seq[Certificate] = {
  val bais = new ByteArrayInputStream(file)
  try{
    CertificateFactory.getInstance("X.509").generateCertificates(bais).toList
  } finally {
  case NonFatal(_) =>
    bais.close()
  // handle error...
  }
}
```
Then we avoid having to read from file again which is an expensive operation

File: conf/application.conf (hunk 4 above)
Comment (sureshhmrc) [inline conf/application.conf]: we don't want to log handshake debug logs which is huge

File: test/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnectorSpec.scala (hunk 5 above)
Comment (sureshhmrc) [inline test/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnectorSpec.scala]: fixed

**Comment by AndyHWChung** [comment]
LGTM

**Pull request merged**